### PR TITLE
Validate stream name

### PR DESCRIPTION
**Pull request by npunniyamoorthy**
Validate fields when streams are created and reconfigured for AirPlay, Spotify, Pandora, LMS, and Internet Radio

### Checklist

* [x] Have you tested your changes and ensured they work?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [x] If applicable, have you updated the CHANGELOG?
* [x] Does your submission pass linting & tests? You can test on localhost using `./scripts/test`

### CODE REVIEW

**Comment by npunniyamoorthy** [comment]
Need to: 
-have validate_stream take in a dictionary instead of multiple arguments. 
-Create a type of exception for streams so we can have consistent error messages

File: amplipi/streams.py [hunk 1]
```diff
@@ -86,6 +86,12 @@ def uuid_gen():
   # Generic UUID in case of failure
   return '39ae35cc-b4c1-444d-b13a-294898d771fa'
 
+class StreamError(Exception):
```
Comment (linknum23) [inline amplipi/streams.py:89]: Would this work better as an InvalidField exception? That way the code could always print the message with the same formatting.

File: amplipi/ctrl.py [hunk 2]
```diff
@@ -984,8 +988,9 @@ def set_stream(self, sid: int, update: models.StreamUpdate) -> ApiResponse:
       stream.reconfig(**changes)
       self._sync_stream_info()
       return ApiResponse.ok()
-    except Exception as exc:
```
[the remark this answers is not in the document]
Comment (npunniyamoorthy) [reply amplipi/ctrl.py:987]: Probably not safe to assume so

File: web/src/pages/Settings/Streams/StreamModal/StreamModal.jsx [hunk 3]
```diff
@@ -133,19 +133,6 @@ InternetRadioSearch.propTypes = {
     onChange: PropTypes.func.isRequired,
 };
 
-function validateInput(streamFields, streamTemplate) {
-    streamTemplate['fields'].filter((f) => f.required).map( (f) => {
```
Comment (linknum23) [inline web/src/pages/Settings/Streams/StreamModal/StreamModal.jsx:137]: Does this move all required field validation to the api?

Comment (npunniyamoorthy) [reply web/src/pages/Settings/Streams/StreamModal/StreamModal.jsx]: yes

**Comment by linknum23** [comment]
As far as I can tell all of the previous review comments have not been addressed. 

In addition since this is an API change we should probably update the stream tests in tests/test_rest.py:
- [ ] Make test_patch_stream_rename to check for success or failure depending on a list of good or bad names. 
- [ ] Make test_create_pandora check for success or failure based on good/bad names
- [ ] Add a similar test_create_X for the other stream names being validated

File: amplipi/ctrl.py [hunk 4]
```diff
@@ -83,6 +88,11 @@ def error(msg: str):
     """ create an error response """
     return ApiResponse(ApiCode.ERROR, msg)
 
+  @staticmethod
```
Comment (linknum23) [inline amplipi/ctrl.py:91]: :heart:

File: docs/amplipi_api.yaml [hunk 5]
```diff
@@ -1,12047 +0,0 @@
-openapi: 3.0.2
```
Comment (linknum23) [inline docs/amplipi_api.yaml:1]: Why was this deleted?

File: amplipi/streams.py [hunk 6]
```diff
@@ -1167,6 +1200,14 @@ def send_cmd(self, cmd):
     except Exception:
       pass
 
+  def validate_stream(self, **kwargs):
+    url_regex = r'^https?://[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)$'
+    logo_regex = r'^https?://[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)\.(?:jpg|gif|png)$'
```
Comment (linknum23) [inline amplipi/streams.py:1205]: We probably don't need to specify the type of image here. Web browsers support a wide variety of image formats and I assume that they don't require an extension.

File: amplipi/streams.py [hunk 7]
```diff
@@ -685,6 +704,11 @@ def send_cmd(self, cmd):
     except Exception as e:
       raise Exception(f"Error sending command {cmd}: {e}") from e
 
+  def validate_stream(self, **kwargs):
+    regex = r"^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$"
```
Comment (linknum23) [inline amplipi/streams.py:708]: I'm having a hard time digesting this regex. This is the one that detects a hostname-like name right?

File: amplipi/streams.py (hunk 6 above)
Comment (linknum23) [reply amplipi/streams.py:1205]: I assume both fields could just use url_regex right?

File: web/src/pages/Settings/Streams/StreamModal/StreamModal.jsx [hunk 8]
```diff
@@ -151,33 +155,31 @@ const InternetRadioSearch = ({ onChange }) => {
...
     ); // set streamFields to copy of stream
 
     const [errorMessage, setErrorMessage] = React.useState("");
 
+    const [errorField, setErrorField] = React.useState("");
+
     const streamTemplate = StreamTemplates.filter(
         (t) => t.type === stream.type
     )[0];
+    // HACK slight hack to ensure name is always present
```
[the remark this answers is not in the document]
Comment (linknum23) [reply web/src/pages/Settings/Streams/StreamModal/StreamModal.jsx:174]: This is because the API  Stream's base class requires a name right?

**Review by rtertiaer — CHANGES_REQUESTED** [review]

File: web/src/pages/Settings/Streams/StreamModal/StreamModal.jsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (klay2000) [reply web/src/pages/Settings/Streams/StreamModal/StreamModal.jsx:174]: kinda, I'm gonna clean this up, it's a little overengineered

File: amplipi/streams.py (hunk 7 above)
Comment (linknum23) [reply amplipi/streams.py:708]: once this regex is cleaned up a potential better name for the `regex` variable might be `HOSTNAME_LIKE`

**Review by rtertiaer — APPROVED** [review]
:+1:

File: amplipi/ctrl.py [hunk 9]
```diff
@@ -69,9 +70,13 @@ class ApiCode(Enum):
...
   """ Ctrl Api Response object """
 
-  def __init__(self, code: ApiCode, msg: str = ''):
+  def __init__(self, code: ApiCode, msg: str = '', data: Optional[Dict] = None):
     self.code = code
     self.msg = msg
+    if data:
+      self.data = data
+    else:
+      self.data = {}
```
[the remark this answers is not in the document]
Comment (linknum23) [reply amplipi/ctrl.py:79]: I prefer the original to avoid pitfalls with mutable default arguments. Your suggestion should produce a linter warning.

File: amplipi/streams.py [hunk 10]
```diff
@@ -144,7 +156,7 @@ def restart(self):
       self.send_cmd('stop')
     except:
       logger.info(f'Stream {self.name} does not have a stop response')
-    last_src = self.src # Disconnect sets self.src to none, so temp variable used to keep track
+    last_src = self.src  # Disconnect sets self.src to none, so temp variable used to keep track
```
Comment (linknum23) [inline amplipi/streams.py:159]: Whats going on with the formatter here? I'm seeing a fair amount of lines that are just spacing changes.

Comment (linknum23) [reply amplipi/streams.py:159]: Looks like you are using vscode's autoformatting. We should probably do this to our entire codebase. PR made for this here: https://github.com/micro-nova/AmpliPi/pull/773

File: amplipi/ctrl.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (rtertiaer) [reply amplipi/ctrl.py:79]: TIL. surprised this is the first time I've seen this :woozy_face:
